tools/generators/files_and_groups/src/Generator/PathTree.swift

```diff
@@ -0,0 +1,131 @@
+                    children: []
+                ))
+        }
+
+        for componentCount in (1...nodesByComponentCount.keys.max()!)
+            .reversed()
+        {
+            let nodes = nodesByComponentCount
+                .removeValue(forKey: componentCount)!
+
+            let sortedNodes = nodes.sorted { lhs, rhs in
+                // Already bucketed to have the same component count, so we
+                // don't sort on count first
+
+                for i in lhs.components.indices {
+                   let lhsComponent = lhs.components[i]
+                   let rhsComponent = rhs.components[i]
+                   guard lhsComponent == rhsComponent else {
+                       // We properly sort in `toFileElements()`, so we do a
+                       // simple version here
+                       return lhsComponent < rhsComponent
+                   }
+                }
+
+                guard lhs.isFolder == rhs.isFolder else {
+                    return lhs.isFolder
+                }
+
+                return false
+            }
+
+            // Create parent nodes
+
+            let firstNode = sortedNodes[0]
+            var collectingParentComponents = firstNode.components.dropLast(1)
+            var collectingParentChildren: [PathTreeNode] = []
+            var nodesForNextComponentCount: [PathTreeNodeToVisit] = []
+
+            for node in sortedNodes {
+                let parentComponents = node.components.dropLast(1)
+                if parentComponents != collectingParentComponents {
+                    nodesForNextComponentCount.append(
+                        PathTreeNodeToVisit(
+                            components: Array(collectingParentComponents),
+                            children: collectingParentChildren
+                        )
+                    )
+
+                    collectingParentComponents = parentComponents
+                    collectingParentChildren = []
+                }
+
+                collectingParentChildren.append(
+                    PathTreeNode(
+                        name: String(node.components.last!),
+                        isFolder: node.isFolder,
+                        children: node.children
+                    )
+                )
+            }
+
+            guard componentCount != 1 else {
+                // Root node
+                return PathTreeNode(
+                    name: "",
+                    children: collectingParentChildren
+                )
+            }
+
+            // Last node
+            nodesForNextComponentCount.append(
+                PathTreeNodeToVisit(
+                    components: Array(collectingParentComponents),
+                    children: collectingParentChildren
+                )
+            )
+
+            nodesByComponentCount[componentCount - 1, default: []]
+                .append(contentsOf: nodesForNextComponentCount)
+        }
+
+        fatalError("Unreachable")
```

Comment:
If we ever hit this it would be good to see a more descriptive msg (ignore if it's WIP code)

Reply:
It's unreachable because of a check within the `while` loop. I think I'll adjust the message to be `""` instead, and put more in a comment above it. Otherwise the string takes up space in the binary.

Reply:
>Otherwise the string takes up space in the binary.

Good point